Nrf9160 soc

ioannisg opened the pull request:
This PR contributes the nRF9160 SOC definition and the basic init configuration

ioannisg committed:
soc: nordic: add Kconfig structure for nRF9160 SOC definition

This commit introduces the Kconfig file structure for nrf9160
SoC in the Zephyr tree.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
soc: nordic: add nRF9160 soc and mpu init configuration

This commit adds the nRF9160 fixup and the source files
for SOC and MPU boot configuration.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
soc: arm: introduce Kconfig option for Nordic IDAU

This commit introduces a new Kconfig symbol to signify
that a CPU implements the Nordic-specific security
attribution unit. The symbol is meant to be selected
by Nordic SoCs that implement the System Protection
Unit (SPU) peripheral.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on soc/arm/Kconfig:
```diff
@@ -20,3 +20,12 @@ config CPU_HAS_ARM_SAU
 	help
 	  MCU implements the ARM Security Attribution Unit (SAU).
+
+config CPU_HAS_NRF_IDAU
```
FYI @agross-linaro : this option represents the Nordic IDAU.

pizi-nordic commented on soc/arm/Kconfig:
```diff
@@ -20,3 +20,12 @@ config CPU_HAS_ARM_SAU
+	depends on SOC_SERIES_NRF91X
+	select CPU_HAS_TEE
+	help
+	  MCU implements the nRF (Implementation-defined)
```
`(Implementation defined)` after nRF does not look well.
What about "MCU implements Nordic flavour of Security... (IDAU)."

ioannisg replied:
new text:

> MCU implements the nRF (vendor-specific) Security Attribution Unit.
 (IDAU: "Implementation-Defined Attribution Unit", in accordance with
  ARM terminology).

Hope this clarifies it better, @pizi-nordic

pizi-nordic commented on soc/arm/nordic_nrf/nrf91/Kconfig.defconfig.nrf9160_SICA:
```diff
@@ -0,0 +1,20 @@
+
+config SOC
+	string
+	default "nRF9160_SICA"
```
Why "SICA"?

ioannisg replied:
@pizi-nordic  This is the code-name for current flavor of nrf9160

ioannisg replied:
See the product options of nRF9160: https://www.nordicsemi.com/DocLib/Content/Product_Spec/nRF9160/latest/nRF9160_html5_keyfeatures

pizi-nordic commented on soc/arm/nordic_nrf/nrf91/Kconfig.series:
```diff
@@ -0,0 +1,26 @@
+	select SOC_FAMILY_NRF
+	select NRF_RTC_TIMER
+	select CLOCK_CONTROL
+	select CLOCK_CONTROL_NRF5
```
Why you need CLOCK_CONTROL when you are requesting CLOCK_CONTROL_NRF5?
Is there a missing requirement in CLOCK_CONTROL_NRF5?

ioannisg replied:
nRF51 and nRF52 SOCs (and simulated nRF boards) select both Kconfig options - this is perhaps a bit problematic, but cannot be solved under this PR. We should need a separate pull request for that. Would you be OK with this? I can track it under #12199

ioannisg replied:
@pizi-nordic @agross-linaro I did some investigation here:
- In some SOC/SOC_SERIES (e.g. NXP),the vendor-specific option is defined in the "defconfig" file, and set to "default=y" if CLOCK_CONTROL is selected already (at SOC level)
- In several SOC/SOC_SERIES (e.g. STM32) CLOCK_CONTROL_<vendor-specific> is _selected_ conditionally,if CLOCK_CONTROL is selected (normally) at the BOARD level.
For example `select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL`
- For Nordic nRFx, CLOCK_CONTROL is directly selected at the SOC_SERIES_NRFxx. That is why we only need:  `select CLOCK_CONTROL_NRF5`, i.e. we can skip the conditional (if CLOCK_CONTROL)

Conclusions:
1. It is the other way @agross-linaro , i.e. CLOCK_CONTROL selects the vendor-specific option
2. I can make this look like the STM32 case, i.e. `select CLOCK_CONTROL_NRF5 if CLOCK_CONTROL` but it is not really needed since CLOCK_CONTROL is selected under the same symbol. So I would rather leave it as is.


---

We _do_ have an issue with naming, here, since CLOCK_CONTROL_NRF5 is "used" in nRF91 series. I think we can  simply rename CLOCK_CONTRO_NRF5 --> CLOCK_CONTROL_NRF. I think I will open a separate PR for this.

carlescufi replied:
>We do have an issue with naming, here, since CLOCK_CONTROL_NRF5 is "used" in nRF91 series. I think we can simply rename CLOCK_CONTRO_NRF5 --> CLOCK_CONTROL_NRF. I think I will open a separate PR for this.

Yes, this is the correct approach in my opinion. Let's go for this.

ioannisg replied:
@agross-linaro the CLOCK_CONTROL_NRF5 has been renamed to CLOCK_CONTROL_NRF, I suppose this issue could be resolved?

pizi-nordic commented on soc/arm/nordic_nrf/nrf91/dts_fixup.h:
```diff
@@ -0,0 +1,39 @@
+
+/* SoC level DTS fixup file */
+
+#define DT_NUM_IRQ_PRIO_BITS        DT_ARM_V8M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
```
It looks that there is some alignment problem on this line.

ioannisg replied:
Thanks, fixing it.

pizi-nordic commented on soc/arm/nordic_nrf/nrf91/mpu_regions.c:
```diff
@@ -0,0 +1,33 @@
+#include <arch/arm/cortex_m/mpu/arm_mpu.h>
+
+
+#define XICR_BASE	0x10000000
```
Do we have any header with these value defined?
Why we are not fetching these addresses from dts?
Why we have defines which are not used?

ioannisg replied:
I think this can be cleaned-up. These defines are in the .c file, thus, cannot be used outside this source file. I added them to be consistent with nRF52 but I see that they are neither used in nRF52. Similar defines are added in some NXP and ST SOCs. I suggest open a PR to clean them up, since they are not used.

ioannisg replied:
Removed, already, from this file.

pizi-nordic commented on soc/arm/nordic_nrf/nrf91/soc.c:
```diff
@@ -0,0 +1,74 @@
+
+	ARG_UNUSED(arg);
+
+	key = irq_lock();
```
Are we really expect and an interrupt on SoC init level?

ioannisg replied:
I agree that it can happen.

agross-oss reviewed (commented):
as a whole it looks ok.  I'd like to see the CLOCK_CONTROL fixed.  Aside from that, the rest of the comments are nits.

anangl commented on soc/arm/nordic_nrf/nrf91/soc.c:
```diff
@@ -0,0 +1,74 @@
+#endif
+
+#include <nrf.h>
+#include <hal/nrf_power.h>
```
These two inclusions do not seem to be needed (`<nrf.h>` is included by `<nrfx.h>`, and POWER HAL is not used here).

ioannisg replied:
OK, removed.
Note that the inclusion of nrf.h is present in the nrf51,51 soc.c files, too, so we need to clean this up from those, too.

anangl commented on soc/arm/nordic_nrf/nrf91/soc.h:
```diff
@@ -0,0 +1,27 @@
+
+#ifndef _ASMLANGUAGE
+
+#include <nrf_common.h>
```
This is needless.

ioannisg replied:
Removed.
Again, note that this inclusion is present in nrf51, and 52 soc.h headers.

anangl commented on soc/arm/nordic_nrf/nrf91/dts_fixup.h:
```diff
@@ -0,0 +1,39 @@
+
+#define DT_NUM_IRQ_PRIO_BITS        DT_ARM_V8M_NVIC_E000E100_ARM_NUM_IRQ_PRIORITY_BITS
+
+#define DT_ADC_0_NAME                   DT_NORDIC_NRF_SAADC_ADC_0_LABE
```
One character (`L`) at the end of the line got eaten somehow.

ioannisg replied:
Fixing it, thanks.

lemrey commented on soc/arm/nordic_nrf/nrf91/soc.c:
```diff
@@ -0,0 +1,74 @@
+#define NMI_INIT()
+#endif
+
+#if defined(CONFIG_SOC_NRF9160)
```
Why is this necessary?

ioannisg replied:
This keeps consistency with nRF52 soc.c, where there are multiple SOC variants. In the future we might have multiple nRF91 SOC variants, as well.

lemrey commented on soc/arm/nordic_nrf/nrf91/soc.c:
```diff
@@ -0,0 +1,74 @@
+#include <nrf.h>
+#include <hal/nrf_power.h>
+
+#define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
```
I'd suggest to pass this to the `LOG_MODULE_REGISTER` macro directly.

ioannisg replied:
Is there any reason to do this differently from what we do in nRF51 and nRF52 soc.c?

pizi-nordic replied:
The logger was (not so) recently updated to support simpler registration scheme. We should promote it :)

lemrey commented on soc/arm/nordic_nrf/nrf91/soc.c:
```diff
@@ -0,0 +1,74 @@
+#include <hal/nrf_power.h>
+
+#define LOG_LEVEL CONFIG_SOC_LOG_LEVEL
+#include <logging/log.h>
```
I think this can be moved next to the other inlcudes.

ioannisg replied:
Will do. Note that the same needs to be done in nrf51 and 52 soc. files.

lemrey commented on soc/arm/nordic_nrf/nrf91/dts_fixup.h:
```diff
@@ -0,0 +1,39 @@
+#define DT_UART_3_NAME			DT_NORDIC_NRF_UARTE_UART_3_LABEL
+
+#define DT_FLASH_DEV_NAME \
+	DT_NRF_NRF91_FLASH_CONTROLLER_FLASH_CONTROLLER_LABEL
```
I am unable to find `DT_NRF_NRF91_FLASH_CONTROLLER_FLASH_CONTROLLER_LABEL` in a `generated_dts_board.h` file for a project built for `nRF9160_pca10090`. Could you double check that this is correct?

ioannisg replied:
This is a nice comment, thanks. So, this has been incorrect in downstream. where the fixup was defining `DT_FLASH_DEV_NAME `as `DT_NRF_NRF52_FLASH_CONTROLLER_4001E000_LABEL`. This is incorrect, not only for the "52" in-fix in the name, but also because of wrong address label - it may be 4001E000 or 5001E000, depending on security domain.

This has been already fixed in upstream, after merging the nrf91 DTS support. Therefore, we will now be able to generate the correct define:  	`DT_NRF_NRF91_FLASH_CONTROLLER_FLASH_CONTROLLER_LABEL `

ioannisg commented on soc/arm/nordic_nrf/nrf91/Kconfig.soc:
```diff
@@ -0,0 +1,75 @@
+	select HAS_HW_NRF_UARTE1
+	select HAS_HW_NRF_UARTE2
+	select HAS_HW_NRF_UARTE3
+	select HAS_HW_NRF_WDT
```
I think so. For now, we keep compliance with nRF52 and 51

ioannisg commented:
@lemrey @pizi-nordic @anangl , I have addressed your comments, please, re-visit this review. 

[only issue remaining is the Ksymbols for CLOCK_CONTROL]

lemrey reviewed (approved):
LGTM

carlescufi reviewed (approved):
LGTM, thanks!

carlescufi commented:
@agross-linaro @galak I think we can merge this one and rework the clock control issue in a follow-up PR that @ioannisg is working on.

ioannisg commented:
> 
> 
> @agross-linaro @galak I think we can merge this one and rework the clock control issue in a follow-up PR that @ioannisg is working on.

@carlescufi @agross-linaro @galak, the refactoring of CLOCK_CONTROL_NRF symbol is already tracked as sub-task in #12199

nashif commented:
What boards use this SoC? Can we submit a user in the same PR?

ioannisg commented:
@galak this is now rebased over the latest changes in nordic dts
